symbolic/src/Data/Macaw/Symbolic.hs

```diff
@@ -185,30 +191,34 @@ import qualified Data.Macaw.Symbolic.MemTraceOps as MT
 class ArchInfo arch where
   archVals :: proxy arch -> Maybe (ArchVals arch)
 
+data LLVMMemory
+
+class IsMemoryModel mem where
+  type MemModelType mem arch :: C.CrucibleType
+  type MemModelConstraint mem sym :: Constraint
+
+instance IsMemoryModel LLVMMemory where
+  type MemModelType LLVMMemory arch = MM.Mem
+  type MemModelConstraint LLVMMemory sym = MM.HasLLVMAnn sym
+
+class IsMemoryModel mem => HasMemoryModel arch mem where
+  -- | This function provides a context with a callback that gives access to the
+  -- set of architecture-specific function evaluators ('MacawArchEvalFn'), which
+  -- is a required argument for 'macawExtensions'.
+  withArchEvalGen
+      :: forall a m sym proxy
+       . (IsSymInterface sym, MemModelConstraint mem sym, MonadIO m)
+      => proxy mem
+      -> sym
+      -> (SB.MacawArchEvalFn sym (MemModelType mem arch) arch -> m a)
+      -> m a
+
 -- | The information to support use of macaw-symbolic for a given architecture
-data ArchVals arch = ArchVals
+data ArchVals arch = HasMemoryModel arch LLVMMemory => ArchVals
```

Comment:
Is the consequence here that the only way we can symbolically simulate machine code is if we use the LLVM memory model?  How painful is it to restructure this to make the memory model a parameter to `ArchVals`?

Reply:
Separately, does this mean to introduce the `HasMemoryModel` constraint you have to pattern match on the `ArchValsValue` constructor?  That seems like I would need a comment every time I did it to explain what was happening.  Why not just have a `withMemoryModel` constructor?